system/core/database.php

```diff
@@ -560,6 +560,44 @@ function serveFile($filename, $root) {
 			die("Invalid filename: $filename");
 		}
 		header('Content-Type: ' . getMimeType($filename));
+	  $filesize = filesize($filename);
+		$shortlen =$filesize-1;
+
+		if(getMimeType($filename)==='audio/mpeg'){
+
+			foreach (getallheaders() as $name => $value) {
+			    if($name === 'Range'){
+						// bytes=0-50
+						$range = substr($value,6); // "bytes=start-end" -> "start-end"
+						$x = explode('-',$range); // "start-end" -> ["start","end"]
+						$start = intval($x[0]);
+						$end = intval($x[1]);
+						http_response_code(206);
+						$bytesize=$end-$start;
+						header( 'Content-Length: ' . (string)($bytesize));
+						header( 'Content-Range: bytes '.$start.'-'.$end.'/'.$filesize);
+						header("Pragma: public");
+				 		header("Expires: 0");
+					  header( 'Accept-Ranges: bytes');
+						header('X-Pad: avoid browser bug');
+					  header('Cache-Control: no-cache');
+						header('Content-Disposition: inline; filename="' . 'test.mp3' . '"');
+
+						$handle = fopen($filename, "r");
+						$contents = fread($handle, $end);
+						fclose($handle);
+						echo substr($contents,$start);
```

Comment:
This is really non-ideal, better to not read everything (see e.g. https://gist.github.com/codler/3906826 or https://www.php.net/manual/en/function.readfile.php#86244).

Even better would be to just tell PHP to handle all the reading/writing (i.e. like `readfile` but for a part of the file), so that not all of the data has to pass through PHP variables, but I couldn't quickly find any way to do that (there is `passthru()`, but that always reads until EOF).

Reply:
Agreed, haven't found a way to do that yet, will do some searching

Reply:
fseek https://stackoverflow.com/a/16741352  but that still needs to pass through a variable

Reply:
Yeah, I suspect that's as good as it gets. Have a look at my two links, which do a block-based approach which is *probably* more efficient that loading *all* data into memory at once with `fgets` like in your link.

Reply:
I now have these two working:
```
fseek($handle,$start,0);
print fread($handle,$bytesize);
```
or

```			
$cur=$begin;
fseek($handle,$begin,0);
while(!feof($handle)&&$cur<$end&&(connection_status()==0)){ 
  print fread($handle,min(1024*16,$end-$cur));
  $cur+=1024*16;
}
```

Do I understand correctly that you think the second one would be better?

Reply:
That's correct.

Though the the direct `print(fread(...))` combination *could* allow PHP to optimize this and never store the result in PHP variable (in which case the first approach would be fine), I'm afraid that it won't and ends up reading everything into memory first.